File: maki-code-index/src/index/elixir.rs
use tree_sitter::Node;

use super::common::{
    LanguageExtractor, Section, SkeletonEntry, compact_ws, find_child, line_range, node_text,
};

pub(crate) struct ElixirExtractor;

const DEF: &str = "def";
const DEFP: &str = "defp";
const DEFMODULE: &str = "defmodule";

const IMPORT_KEYWORDS: &[&str] = &["alias", "import", "require", "use"];

const DOC_ATTRS: &[&str] = &["doc", "moduledoc", "typedoc", "spec"];

impl ElixirExtractor {
    fn call_target<'a>(&self, node: Node, source: &'a [u8]) -> Option<&'a str> {
        if node.kind() != "call" {
            return None;
        }
        let target = node.child_by_field_name("target")?;
        Some(node_text(target, source))
    }

    fn extract_import(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let target_name = self.call_target(node, source)?;
        if !IMPORT_KEYWORDS.contains(&target_name) {
            return None;
        }
        let args = find_child(node, "arguments")?;
        let first_arg = args.child(0)?;
        let module_path = node_text(first_arg, source);
        let segments: Vec<String> = module_path.split('.').map(String::from).collect();
        Some(SkeletonEntry::new_import_with_keyword(
            node,
            Some(target_name.to_string()),
            vec![segments],
        ))
    }

    fn extract_module(&self, node: Node, source: &[u8]) -> Vec<SkeletonEntry> {
        match self.call_target(node, source) {
            Some(DEFMODULE) => {}
            _ => return Vec::new(),
        };
        let args = match find_child(node, "arguments") {
            Some(a) => a,
            None => return Vec::new(),
        };
        let name = match args.child(0) {
            Some(c) => node_text(c, source),
            None => return Vec::new(),
        };
        let mut imports = Vec::new();
        let mut methods = Vec::new();
        if let Some(block) = find_child(node, "do_block") {
            let mut cursor = block.walk();
            for child in block.children(&mut cursor) {
                if let Some(imp) = self.extract_import(child, source) {
                    imports.push(imp);
                    continue;
                }
                if self.is_def_call(child, source).is_some()
                    && let Some(sig) = self.def_sig(child, source)
                {
                    let lr =
                        line_range(child.start_position().row + 1, child.end_position().row + 1);
                    methods.push(compact_ws(&format!("{sig} {lr}")).into_owned());
                }
            }
        }
        imports.push(
            SkeletonEntry::new(Section::Class, node, format!("defmodule {name}"))
                .with_children(methods),
        );
        imports
    }
    fn def_sig(&self, node: Node, source: &[u8]) -> Option<String> {
        let args = find_child(node, "arguments")?;
        let first = args.child(0)?;
        match first.kind() {
            "call" | "identifier" | "binary_operator" => Some(node_text(first, source).to_string()),
            _ => None,
        }
    }

    fn is_def_call<'a>(&self, node: Node, source: &'a [u8]) -> Option<&'a str> {
        let target_name = self.call_target(node, source)?;
        (target_name == DEF || target_name == DEFP).then_some(target_name)
    }

    fn extract_standalone_fn(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        self.is_def_call(node, source)?;
        let sig = self.def_sig(node, source)?;
        Some(SkeletonEntry::new(
            Section::Function,
            node,
            compact_ws(&sig).into_owned(),
        ))
    }

    fn extract_module_attr(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let name = self.attr_name(node, source)?;
        if DOC_ATTRS.contains(&name.as_str()) {
            return None;
        }
        if !name.starts_with(|c: char| c.is_ascii_uppercase()) {
            return None;
        }
        Some(SkeletonEntry::new(
            Section::Constant,
            node,
            format!("@{name}"),
        ))
    }

    fn attr_name(&self, node: Node, source: &[u8]) -> Option<String> {
        if node.kind() != "unary_operator" {
            return None;
        }
        let op = node.child_by_field_name("operator")?;
        if node_text(op, source) != "@" {
            return None;
        }
        let operand = node.child_by_field_name("operand")?;
        match operand.kind() {
            "identifier" | "alias" => Some(node_text(operand, source).to_string()),
            "call" => {
                let target = operand.child_by_field_name("target")?;
                Some(node_text(target, source).to_string())
            }
            _ => None,
        }
    }
}

impl LanguageExtractor for ElixirExtractor {
    fn extract_nodes(&self, node: Node, source: &[u8], _attrs: &[Node]) -> Vec<SkeletonEntry> {
        if node.kind() == "call" {
            if let Some(entry) = self.extract_import(node, source) {
                return vec![entry];
            }
            let module_entries = self.extract_module(node, source);
            if !module_entries.is_empty() {
                return module_entries;
            }
            if let Some(entry) = self.extract_standalone_fn(node, source) {
                return vec![entry];
            }
        }
        if let Some(entry) = self.extract_module_attr(node, source) {
            return vec![entry];
        }
        Vec::new()
    }

    fn is_doc_comment(&self, node: Node, source: &[u8]) -> bool {
        if node.kind() == "comment" {
            return true;
        }
        self.attr_name(node, source)
            .is_some_and(|name| DOC_ATTRS.contains(&name.as_str()))
    }

    fn import_separator(&self) -> &'static str {
        "."
    }
}

File: plugins/index/lang/elixir.lua
return function(U)
  local get_text = U.get_text
  local find_child = U.find_child
  local compact_ws = U.compact_ws
  local format_range = U.format_range
  local line_start = U.line_start
  local line_end = U.line_end
  local new_entry = U.new_entry
  local new_import_entry = U.new_import_entry
  local SECTION = U.SECTION

  local DEF = "def"
  local DEFP = "defp"
  local IMPORT_KEYWORDS = { alias = true, import = true, require = true, use = true }
  local DOC_ATTRS = { doc = true, moduledoc = true, typedoc = true, spec = true }

  local function call_target(node, source)
    if node:type() ~= "call" then return nil end
    local t = node:field("target")[1]
    return t and get_text(t, source) or nil
  end

  local function first_arg(node, source)
    local args = find_child(node, "arguments")
    if not args then return nil end
    for _, child in ipairs(args:children()) do
      local ck = child:type()
      if ck ~= "," and ck ~= "(" and ck ~= ")" then
        return child
      end
    end
    return nil
  end

  local function split_dot(text)
    local parts = {}
    for p in text:gmatch("[^%.]+") do
      parts[#parts + 1] = p
    end
    return parts
  end

  local function extract_import(node, source)
    local target = call_target(node, source)
    if not target or not IMPORT_KEYWORDS[target] then return nil end
    local arg = first_arg(node, source)
    if not arg then return nil end
    local arg_text = get_text(arg, source)
    local paths = { split_dot(arg_text) }
    return new_import_entry(node, paths, target)
  end

  local function def_sig(node, source)
    local args = find_child(node, "arguments")
    if not args then return nil end
    local arg = first_arg(node, source)
    if not arg then return nil end
    local ck = arg:type()
    if ck == "call" or ck == "identifier" or ck == "binary_operator" then
      return get_text(arg, source)
    end
    return nil
  end

  local function is_def_call(node, source)
    local target = call_target(node, source)
    return target == DEF or target == DEFP
  end

  local function extract_standalone_fn(node, source)
    if not is_def_call(node, source) then return nil end
    local sig = def_sig(node, source)
    if not sig then return nil end
    local target = call_target(node, source)
    local prefix = target == DEFP and "defp " or "def "
    return new_entry(SECTION.Function, node, compact_ws(prefix .. sig))
  end

  local function attr_name(node, source)
    if node:type() ~= "unary_operator" then return nil end
    local op = node:field("operator")[1]
    if not op or get_text(op, source) ~= "@" then return nil end
    local operand = node:field("operand")[1]
    if not operand then return nil end
    local ok = operand:type()
    if ok == "identifier" or ok == "alias" then
      return get_text(operand, source)
    elseif ok == "call" then
      return call_target(operand, source)
    end
    return nil
  end

  local function extract_module_attr(node, source)
    local name = attr_name(node, source)
    if not name then return nil end
    if DOC_ATTRS[name] then return nil end
    if not name:match("^[A-Z]") then return nil end
    return new_entry(SECTION.Constant, node, "@" .. name)
  end

  local function extract_module(node, source)
    local target = call_target(node, source)
    if target ~= "defmodule" then return nil end
    local arg = first_arg(node, source)
    if not arg then return nil end
    local name = get_text(arg, source)

    local do_block = find_child(node, "do_block")
    if not do_block then
      return new_entry(SECTION.Class, node, "defmodule " .. name)
    end

    local imports = {}
    local methods = {}

    for _, child in ipairs(do_block:children()) do
      local ck = child:type()
      if ck == "call" then
        local imp = extract_import(child, source)
        if imp then
          imports[#imports + 1] = imp
        elseif is_def_call(child, source) then
          local sig = def_sig(child, source)
          if sig then
            local t = call_target(child, source)
            local prefix = t == DEFP and "defp " or "def "
            local lr = format_range(line_start(child), line_end(child))
            methods[#methods + 1] = compact_ws(prefix .. sig) .. " " .. lr
          end
        end
      end
    end

    local entry = new_entry(SECTION.Class, node, "defmodule " .. name)
    entry.children = methods
    return entry, imports
  end

  return {
    import_separator = ".",

    is_doc_comment = function(node, source)
      local ck = node:type()
      if ck == "comment" then return true end
      if ck == "unary_operator" then
        local name = attr_name(node, source)
        return name ~= nil and DOC_ATTRS[name] == true
      end
      return false
    end,

    extract_nodes = function(node, source, _attrs)
      local ck = node:type()
      if ck == "call" then
        local imp = extract_import(node, source)
        if imp then return { imp } end
        local mod_entry, mod_imports = extract_module(node, source)
        if mod_entry then
          local result = {}
          if mod_imports then
            for _, mi in ipairs(mod_imports) do
              result[#result + 1] = mi
            end
          end
          result[#result + 1] = mod_entry
          return result
        end
        local fn_entry = extract_standalone_fn(node, source)
        if fn_entry then return { fn_entry } end
      elseif ck == "unary_operator" then
        local e = extract_module_attr(node, source)
        if e then return { e } end
      end
      return {}
    end,
  }
end
